src/gt4py/next/errors/formatting.py

```diff
@@ -0,0 +1,99 @@
+#
+# Copyright (c) 2014-2023, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import linecache
+import textwrap
+import traceback
+import types
+from typing import Optional
+
+from gt4py.eve import SourceLocation
+
+
+def get_source_from_location(location: SourceLocation) -> str:
+    if not location.filename:
+        raise FileNotFoundError()
+    source_lines = linecache.getlines(location.filename)
+    if not source_lines:
+        raise FileNotFoundError()
+    start_line = location.line
+    end_line = location.end_line + 1 if location.end_line is not None else start_line + 1
+    relevant_lines = source_lines[(start_line - 1) : (end_line - 1)]
+    return "\n".join(relevant_lines)
+
+
+def format_location(loc: SourceLocation, show_caret: bool = False) -> str:
+    """
+    Format the source file location.
+
+    Args:
+        show_caret (bool): Indicate the position within the source line by placing carets underneath.
+    """
+    filename = loc.filename or "<unknown>"
+    lineno = loc.line
+    loc_str = f'File "{filename}", line {lineno}'
+
+    if show_caret and loc.column is not None:
+        offset = loc.column - 1
+        width = loc.end_column - loc.column if loc.end_column is not None else 1
+        caret_str = "".join([" "] * offset + ["^"] * width)
+    else:
+        caret_str = None
+
+    try:
+        snippet_str = get_source_from_location(loc)
+        if caret_str:
+            snippet_str = f"{snippet_str}{caret_str}"
+        return f"{loc_str}\n{textwrap.indent(snippet_str, '  ')}"
+    except (FileNotFoundError, IndexError):
+        return loc_str
+
+
+def _format_cause(cause: BaseException) -> list[str]:
+    """Format the cause of an exception plus the bridging message to the current exception."""
+    bridging_message = "The above exception was the direct cause of the following exception:"
+    cause_strs = [*traceback.format_exception(cause), "\n", f"{bridging_message}\n\n"]
+    return cause_strs
+
+
+def _format_traceback(tb: types.TracebackType) -> list[str]:
+    """Format the traceback of an exception."""
+    intro_message = "Traceback (most recent call last):"
+    traceback_strs = [
+        f"{intro_message}\n",
+        *traceback.format_tb(tb),
+    ]
+    return traceback_strs
+
+
+def format_compilation_error(
+    type_: type[Exception],
+    message: str,
+    location: Optional[SourceLocation],
+    tb: Optional[types.TracebackType] = None,
+    cause: Optional[BaseException] = None,
+) -> list[str]:
```

Comment:
Why is this a list? Sometimes you join using an empty string sometimes with "\n". Is there a reason I want to be able to choose? A small docstring with a doctest would be useful.

Reply:
It's to have an interface consistent with [traceback.format_exception](https://docs.python.org/3/library/traceback.html#traceback.format_exception). I have no idea why they chose a list of strings.

Reply:
Nevertheless, it should always be joined without "\n", I'm not sure if I got all of them right.